### PR TITLE
Update: Add 'caughtErrors' to rule no-unused-vars (fixes #3837)

### LINKED ISSUE #3837: Option Proposal: for `no-unused-vars`, add `exceptions` and `exceptionsIgnorePattern` options
Following from the discussion in #3543, I propose the addition of options to the `no-unused-vars` rule, allowing us to specify how we want to handle exception variables defined in the `catch()` clause of a try/catch statement.

Currently, the documentation doesn't address this, and the rule allows caught exception variables to remain unused, even when the `"all"` settings are used for the current options (tested via ESLint 1.4.3, though this behavior was also present in previous versions):
```
/*eslint no-unused-vars: [2, {"vars": "all", "args": "all"}]*/
function example(){
  try {
    doSomething();
  } catch (err) { // unused exception variable isn't flagged as an error
    return false;
  }
  return true;
}
```
Therefore, I propose the addition of `exceptions` and `exceptionsIgnorePattern` options, mirroring the functionality of the `vars`/`varsIgnorePattern` and `args`/`argsIgnorePattern` options. Valid settings for `exceptions` would be:

* `none` - the default, matches the current behavior of not reporting any unused exception variables
* `all` - reports all unused exception variables, aside from those matching `exceptionsIgnorePattern`
```
/*eslint no-unused-vars: [2, {"exceptions": "all", "exceptionsIgnorePattern": "^ignore"}]*/
function hypotheticalExample(){
  try {
    doSomething();
  } catch (err) { // unused exception variable IS flagged as an error
    return false;
  }
  try {
    doSomethingElse();
  } catch (ignoreErr) { // unused exception variable matches ignore pattern, is NOT flagged as an error
    return false;
  }
  return true;
}
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26760407-option-proposal-for-no-unused-vars-add-exceptions-and-exceptionsignorepattern-options?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pvamshi**
Adding new options `exceptions` and `exceptionsIgnorePattern` as mentioned in #3837 . Please verify and let me know if anything more needs to be added.
( sorry for duplicate pull request , still not expert at squashing)

### CODE REVIEW

**Comment by nzakas** [comment]
No problem, squashing takes a little getting used to. The code looks good and the docs are great. 

After looking over the original issue, I'm a bit afraid that "exceptions" is the wrong term because we use it in other rules to represent exceptions to the rule rather than thrown errors (see http://eslint.org/docs/rules/no-multi-spaces#exceptions for example). I'm worried this might confuse people. Maybe `errors` or `caughtErrors` would be clearer? @eslint/eslint-team ?

File: docs/rules/no-unused-vars.md [hunk 1]
```diff
@@ -182,6 +182,63 @@ function foo(x, _y) {
 foo();
 ```
 
+### exceptions
+
+The `exceptions` option is used for `catch` block arguments validation.
+
+It has two settings:
+
+* `none` - do not check arguments. This is the default setting.
```
Comment (nzakas) [inline docs/rules/no-unused-vars.md]: Do not check exception objects

File: docs/rules/no-unused-vars.md [hunk 2]
```diff
@@ -182,6 +182,63 @@ function foo(x, _y) {
...
+
+Examples of **correct** code for the `{ "exceptions": "none" }` option:
+
+```js
+/*eslint no-unused-vars: [2, { "exceptions": "none" }]*/
+
+try {
+    //...
+} catch( err ){
+    console.error("exception");
```
Comment (nzakas) [inline docs/rules/no-unused-vars.md]: No spaces around `err` please.

Comment (platinumazure) [inline docs/rules/no-unused-vars.md]: Also should probably have a space after `catch`.

**Comment by pvamshi** [comment]
@nzakas Thanks. Once we finalize the term to be used, I will update the code accordingly along with these review comments. 

On a side note, I learnt javascript from reading "Javascript for Web Developers " 2nd edition around 5 years back. Great feeling to get code reviewed positively from you :relieved:

File: docs/rules/no-unused-vars.md (hunk 2 above)
Comment (kaicataldo) [inline docs/rules/no-unused-vars.md]: @platinumazure Just to clarify, did you mean `catch (){` or `catch() {`? I'm assuming you meant the latter, but could see the suggestion being taken either way  :)

**Comment by kaicataldo** [comment]
@pvamshi  Great work on this PR! Agreed about using `errors` or `caughtErrors` over `exceptions`. I have a slight preference to `errors` just because `caughtErrorsIgnorePattern` is kind of long, though I can see the argument for `caughtErrors` since I think it makes it clearer that this change is specifically scoped to `catch` statements.

File: docs/rules/no-unused-vars.md (hunk 2 above)
Comment (pvamshi) [inline docs/rules/no-unused-vars.md]: I will take the later :grinning:

**Comment by pvamshi** [comment]
I will go ahead with `errors` then . If someone has an objection , we can take it as a bug. I have been waiting for a reply on this PR for long now :worried:

File: docs/rules/no-unused-vars.md (hunk 2 above)
Comment (platinumazure) [inline docs/rules/no-unused-vars.md]: I meant both before and after the parentheses, as well as no space inside the parens. Sorry about that. `catch (err) {` is correct.

**Comment by kaicataldo** [comment]
Thanks for sticking with it!  Once you make the suggested changes we'll just need  one final stamp of approval from @nzakas and @platinumazure.

File: docs/rules/no-unused-vars.md (hunk 2 above)
Comment (kaicataldo) [inline docs/rules/no-unused-vars.md]: Great, thanks for clarifying 👍

**Comment by nzakas** [comment]
Hmm, I think we need to go with `caughtErrors` because `errors` doesn't really indicate where this applies.

Also, can you update the commit summary to begin with "Update:"? (This is an enhancement rather than a bug fix.)

**Comment by pvamshi** [comment]
Sure , will make the changes .

**Comment by pvamshi** [comment]
Done with the changes , looks good for me . Please verify if I missed out any detail.

**Comment by nzakas** [comment]
This looks great, thanks and nice to meet you!

**Pull request merged**